WalletWasabi/Crypto/ZeroKnowledge/NonInteractive/FiatShamirTransform/Knowledge.cs

```diff
@@ -0,0 +1,78 @@
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge.NonInteractive.FiatShamirTransform
+{
+	public class Knowledge : IKnowledge
+	{
+		// Although in principle this could be an interface type, since only one
+		// type of statement is needed for our proofs this is specialized to only
+		// that type.
+		public Knowledge(LinearRelation.Knowledge knowledge)
+		{
+			LinearRelation = Guard.NotNull(nameof(knowledge), knowledge);
+		}
+
+		public LinearRelation.Knowledge LinearRelation { get; }
+
+		public IStatement ToStatement()
+		{
+			return new FiatShamirTransform.Statement(LinearRelation.Statement);
+		}
+
+		public ProverCommitToNonces CommitToStatements(Transcript transcript)
+		{
+			// Before anything else all components in a compound proof commit to the
```

Comment:
Is a verb missing?

Reply:
I don't think so. If there's an AND proof containing this proof, the first thing that happens is that all of them are declared to the transcript. This is because the nonce generation must depend on everything. After all that is finished, the actual Sigma protocol can begin, by using effectively a hash of all of the above as part of the generation of random commitments.